common/http_test.go

```diff
@@ -0,0 +1,171 @@
+	}
+
+	for _, code := range codes {
+		w := httptest.NewRecorder()
+		common.WriteCode(w, "foo", code)
+		assert.Equal(t, w.Code, code, "expecting the same code recorded")
+	}
+}
+
+type saver struct {
+	Foo int    `json:"foo" validate:"min=10"`
+	Bar string `json:"bar"`
+}
+
+type rcloser struct {
+	closeOnce sync.Once
+	body      *bytes.Buffer
+	closeChan chan bool
+}
+
+var errAlreadyClosed = errors.New("already closed")
+
+func (rc *rcloser) Close() error {
+	var err = errAlreadyClosed
+	rc.closeOnce.Do(func() {
+		err = nil
+		rc.closeChan <- true
+		close(rc.closeChan)
+	})
+	return err
+}
+
+func (rc *rcloser) Read(b []byte) (int, error) {
+	return rc.body.Read(b)
+}
+
+var _ io.ReadCloser = (*rcloser)(nil)
+
+func makeReq(strBody string) (*http.Request, <-chan bool) {
+	closeChan := make(chan bool, 1)
+	buf := new(bytes.Buffer)
+	buf.Write([]byte(strBody))
+	req := &http.Request{
+		Header: make(http.Header),
+		Body:   &rcloser{body: buf, closeChan: closeChan},
+	}
+	return req, closeChan
+}
+
+func TestParseRequestJSON(t *testing.T) {
+	tests := [...]struct {
+		body    string
+		want    interface{}
+		wantErr bool
+	}{
+		0: {wantErr: true, body: ``},
+		1: {body: `{}`},
+		2: {body: `{"foo": 2}`}, // Not that the validate tags don't matter here since we are just parsing
+		3: {body: `{"foo": "abcd"}`, wantErr: true},
+	}
+
+	for i, tt := range tests {
+		req, closeChan := makeReq(tt.body)
+		sav := new(saver)
+		err := common.ParseRequestJSON(req, sav)
+		if tt.wantErr {
+			assert.NotEqual(t, err, nil, fmt.Sprintf("#%d: want non-nil error", i))
+			continue
+		}
+		assert.Equal(t, err, nil, fmt.Sprintf("#%d: want nil error", i))
+		wasClosed := <-closeChan
+		assert.Equal(t, wasClosed, true, fmt.Sprintf("#%d: should have invoked close", i))
+	}
+}
+
+func TestFparseJSON(t *testing.T) {
+	r1 := strings.NewReader(`{"foo": 1}`)
+	sav := new(saver)
+	require.Equal(t, common.FparseJSON(r1, sav), nil, "expecting successful parsing")
+	r2 := strings.NewReader(`{"bar": "blockchain"}`)
+	require.Equal(t, common.FparseJSON(r2, sav), nil, "expecting successful parsing")
+	require.Equal(t, reflect.DeepEqual(sav, &saver{Foo: 1, Bar: "blockchain"}), true, "should have parsed both")
+}
+
+func TestFparseAndValidateJSON(t *testing.T) {
+	r1 := strings.NewReader(`{"foo": 1}`)
+	sav := new(saver)
+	require.NotEqual(t, common.FparseAndValidateJSON(r1, sav), nil, "expecting validation to fail")
+	r1 = strings.NewReader(`{"foo": 100}`)
+	require.Equal(t, common.FparseJSON(r1, sav), nil, "expecting successful parsing")
+	r2 := strings.NewReader(`{"bar": "blockchain"}`)
+	require.Equal(t, common.FparseAndValidateJSON(r2, sav), nil, "expecting successful parsing")
+	require.Equal(t, reflect.DeepEqual(sav, &saver{Foo: 100, Bar: "blockchain"}), true, "should have parsed both")
+}
+
+func TestParseAndValidateRequestJSON(t *testing.T) {
+	tests := [...]struct {
+		body    string
+		want    interface{}
+		wantErr bool
+	}{
+		0: {wantErr: true, body: ``},
```

Comment:
Same comment as above: 
want is always nil? shouldn't you provide a `new(saver)` or something like that.  Otherwise, of course they all error. It would be good to see some positive results as well.  Feel these tests were 80% finished.